doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,264 @@
+  play to avoid re-writing all the data to s3 on each retry.
+- On retry, there's likely a very important performance special case to check if
+  any of the inputs to `update_fn` have changed since the last read_ts and skip
+  recomputing update_fn if not.
+  - Note that with this optimization, this does exactly the same thing as
+    "write-only single-table txn".
+
+The rest of this document concerns the implementation of the durability
+component. Additional detail on the timestamp selection is left for a separate
+design document.
+
+## Detailed description
+
+WIP initially just sketching these out while we develop an opinion on which one
+to go with
+
+General notes:
+- There's a fundamental tension between retries/contention and availability
+  here. At the end of the day, if there's a write txn in some other process
+  that's trying to commit at a given timestamp and it overlaps with what you're
+  doing, you either have to force it to retry or wait for it. (Which latter
+  might never complete b/c distributed systems.) As a result, we play tricks
+  with minimizing the cost of retries, tighter contention surface areas, and
+  pushing retries off of important and/or costly txns.
+
+### Option 1: Put all tables in a single persist shard
+
+- Model each update in the shard as something like `(GlobalId, Row)`
+- Reads would then pull out the table(s) they are interested in, likely
+  optimized in some way by pushdown
+- Pro: Can be implemented entirely on top of persist.
+- Pro: Atomic multi-table writes are trivial, as is advancing the upper of all
+  tables at once.
+- Con: Unintuitive. Nothing else works this way.
+- Con: There are likely to be "gotchas" from putting small+large, low+high
+  throughput tables all in one shard. E.g. when individual retractions from some
+  table consolidate out would be dependent on all the other tables
+
+### Option 2: Put all tables in a single persist "state"
+
+- Currently, persist has a 1:1 relationship between shards and the metadata
+  (batches, capabilities, etc) kept by the state machine. Instead, make the
+  metadata a map of `ShardId -> ShardMetadata`.
+- Pro: Atomic multi-table writes are trivial, as is advancing the upper of all
+  tables at once.
+- Con: Greatly adds complexity to the internals of persist, which is already
+  complex.
+- Con: Greatly complicates the API of persist.
+- Con: Membership changes (the mapping between a shard and which state it lives
+  in) are tricky.
+- Con: Significant risk of destabilizing persist, which has to work perfectly
+  for basically anything in Materialize to work.
+- Con: Reading or writing a single table would require loading the metadata of
+  all co-located shards, which scales arbitrarily large with the number of
+  tables. This likely implies an upper bound on the number of tables we can
+  support.
+- Con: Also likely includes a rewrite of how persist monitoring and debugging
+  tools work.
+
+### Option 3: Table remap shard
```

Comment:
If you need some design inspiration, the transaction protocol from this paper is fairly similar: http://www.vldb.org/pvldb/vol10/p685-zamanian.pdf/.

They talk a lot about using RDMA, which is not relevant here, but at a very high level it's similar in the sense that they:
1. Write some data down with a timestamp.
2. Mark that data committed by updating a map of data -> committed timestamp.

We read this in the paper reading group and I can't remember if you were there that week. If so, then ignore this.